src/v/resource_mgmt/storage.cc

```diff
@@ -218,6 +218,141 @@ eviction_policy::partition* eviction_policy::schedule::current() {
     return &shards[shard_idx].partitions[partition_idx];
 }
 
+ss::future<eviction_policy::schedule> eviction_policy::create_new_schedule() {
+    auto shards = co_await _pm->map([this](auto& /* ignored */) {
+        /*
+         * the partition_manager reference is ignored since
+         * collect_reclaimable_offsets already has access to a
+         * sharded<partition_manager>.
+         */
+        return collect_reclaimable_offsets().then([](auto partitions) {
+            return shard_partitions{
+              .shard = ss::this_shard_id(),
+              .partitions = std::move(partitions),
+            };
+        });
+    });
+
+    // compute the size of the schedule
+    const auto size = std::accumulate(
+      shards.cbegin(),
+      shards.cend(),
+      size_t{0},
+      [](size_t acc, const shard_partitions& shard) {
+          return acc + shard.partitions.size();
+      });
+
+    vlog(rlog.debug, "Created new eviction schedule with {} partitions", size);
+    co_return schedule(std::move(shards), size);
+}
+
+ss::future<fragmented_vector<eviction_policy::partition>>
+eviction_policy::collect_reclaimable_offsets() {
+    /*
+     * build a lightweight copy to avoid invalidations during iteration
+     */
+    fragmented_vector<ss::lw_shared_ptr<cluster::partition>> partitions;
+    for (const auto& p : _pm->local().partitions()) {
+        if (!p.second->remote_partition()) {
+            continue;
+        }
+        partitions.push_back(p.second);
+    }
+
+    /*
+     * retention settings mirror settings found in housekeeping()
+     */
+    const auto collection_threshold = [this] {
+        const auto& lm = _storage->local().log_mgr();
+        if (!lm.config().delete_retention().has_value()) {
+            return model::timestamp(0);
+        }
+        const auto now = model::timestamp::now().value();
+        const auto retention = lm.config().delete_retention().value().count();
+        return model::timestamp(now - retention);
+    };
+
+    gc_config cfg(
+      collection_threshold(),
+      _storage->local().log_mgr().config().retention_bytes());
+
+    /*
+     * in smallish batches partitions are queried for their reclaimable
+     * segments. all of this information is bundled up and returned.
+     */
+    fragmented_vector<partition> res;
+    co_await ss::max_concurrent_for_each(
+      partitions.begin(), partitions.end(), 20, [&res, cfg](const auto& p) {
+          auto log = dynamic_cast<storage::disk_log_impl*>(p->log().get_impl());
+          auto gate = log->gate().hold(); // protect against log deletes
```

Comment:
if the gate is closed should we be catching gate closed, and could `get_relaimable_offsets` invoke gate->hold()?

Reply:
i think you are right on both points. definitely the first, but i recall an issue with the second. i'll try again to move it into get_reclaimable_offsets.

Reply:
good catch. the gate was already held in `get_relaimable_offsets` and we don't need it here--it was a hold over from the previous greedy version that i copied some code from.

also catch and ignore gate closed exception and other exceptions (with logging) so we don't hold up reclaim because one partition has some junky segment files or something.